### ISSUE TITLE
Spurious failure of the vio2vo test

### ISSUE THREAD

**Issue by ppedrot** [coq/coq#7713]
In two unrelated PRs (#7679 and #7004), the vio2vo test from the Coq test-suite failed spuriously.

Links to the failures:
- https://gitlab.com/coq/coq/-/jobs/72408667
- https://gitlab.com/coq/coq/-/jobs/72483736

**Comment by Zimmi48** [comment]
New occurrence:

- https://gitlab.com/coq/coq/-/jobs/72821129

Looks like it has suddenly become very frequent.

cc @gares

**Comment by maximedenes** [comment]
Why don't we see any output in the tests?

**Comment by gares** [comment]
I'm sorry but so far many seemingly STM related tests  were observed to fail randomly.
As far as I know it has never been the fault of Coq, but rather the VMs and the optimistic value of the argument of -j that made it fail.

I've no clue what you are doing on CI. Does it fail on your PC?

**Comment by Zimmi48** [comment]
> Does it fail on your PC?

Good question. If I see one of these tests fail on my PC I'll report it immediately with the log.

**Comment by Zimmi48** [comment]
That being said, I run the test-suite locally much less frequently than I check CI results.

**Comment by SkySkimmer** [comment]
>Why don't we see any output in the tests?

What do you mean?

**Comment by Zimmi48** [comment]
Indeed, we're just stupid and we forgot that GitLab upload artifacts for failed tests: https://gitlab.com/coq/coq/-/jobs/72821129/artifacts/raw/test-suite/logs/coq-makefile/vio2vo.log

**Comment by Zimmi48** [comment]
The log file I linked to is the following BTW:

```
make[1]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
make[1]: Entering directory '/builds/coq/coq/test-suite/coq-makefile/vio2vo/_test'
COQDEP VFILES
CAMLDEP src/test_aux.mli
CAMLDEP src/test.mli
COQDEP src/test_plugin.mlpack
CAMLDEP src/test_aux.ml
CAMLDEP -pp src/test.ml4
make[1]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
CAMLC -c src/test_aux.mli
CAMLOPT -c -for-pack Test_plugin src/test_aux.ml
CAMLC -c src/test.mli
CAMLOPT -pp -c -for-pack Test_plugin src/test.ml4
CAMLOPT -pack -o src/test_plugin.cmx
CAMLOPT -a -o src/test_plugin.cmxa
CAMLOPT -shared -o src/test_plugin.cmxs
COQC -quick theories/test.v
make[1]: Leaving directory '/builds/coq/coq/test-suite/coq-makefile/vio2vo/_test'
make[1]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
make[1]: Entering directory '/builds/coq/coq/test-suite/coq-makefile/vio2vo/_test'
"/builds/coq/coq/_install_ci/bin/coqc"  -q    -I src  -R src test -R theories test \
	-schedule-vio2vo 2 theories/test.vio
Error:
Anomaly
"Uncaught exception Failure("The spawned process did not connect back in 10.0s")."
Please report at http://coq.inria.fr/bugs/.

Makefile:398: recipe for target 'vio2vo' failed
make[1]: *** [vio2vo] Error 129
make[1]: Leaving directory '/builds/coq/coq/test-suite/coq-makefile/vio2vo/_test'
==========> FAILURE <==========
    coq-makefile/vio2vo/run.sh...Error!
```

**Comment by Zimmi48** [comment]
So the problem becomes clear: it is the same one as in #5983. But the timeout bump which allowed to avoid problems in AppVeyor was apparently not enough here. Given that these kind of problems only appear on CI VMs, maybe we should consider relaunching these tests several times instead.

**Comment by gares** [comment]
There is something fishy, 10 seconds even in a VM is really a long time.
Given that today the problem seems reproducible I suggest that we try to set `accept_timeout` to `(-1.0)` (no timeout) and see if the problem goes away. If so, then these VMs sucks, but we have a fix (at least on the VMs, we disable the timeout). If the problems stays, then there must be a race somewhere.

**Comment by gares** [comment]
I think you can make a PR and open an issue to revert it before the release.

**Comment by Zimmi48** [comment]
I just ran the test-suite on master today and I had this failure:

```
make[3]: Entering directory '/home/theo/coq/test-suite/coq-makefile/quick2vo/_test'
make[3]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
Makefile:423: warning: undefined variable 'TGTS'
Makefile:655: warning: undefined variable 'src/test_plugin_MLPACK_DEPENDENCIES'
COQDEP VFILES
CAMLDEP src/test_aux.mli
CAMLDEP src/test.mli
COQDEP src/test_plugin.mlpack
CAMLDEP src/test_aux.ml
CAMLDEP -pp src/test.ml4
make[3]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
Makefile:423: warning: undefined variable 'TGTS'
Makefile:655: warning: undefined variable 'src/test_plugin_MLPACK_DEPENDENCIES'
make[4]: Entering directory '/home/theo/coq/test-suite/coq-makefile/quick2vo/_test'
Makefile:423: warning: undefined variable 'TGTS'
Makefile:655: warning: undefined variable 'src/test_plugin_MLPACK_DEPENDENCIES'
CAMLC -c src/test_aux.mli
CAMLC -c src/test.mli
CAMLOPT -c -for-pack Test_plugin src/test_aux.ml
CAMLOPT -pp -c -for-pack Test_plugin src/test.ml4
CAMLOPT -pack -o src/test_plugin.cmx
CAMLOPT -a -o src/test_plugin.cmxa
CAMLOPT -shared -o src/test_plugin.cmxs
COQC -quick theories/test.v
make[4]: Leaving directory '/home/theo/coq/test-suite/coq-makefile/quick2vo/_test'
VIO2VO: theories/test.vio 
Error:
Anomaly
"Uncaught exception Failure("The spawned process did not connect back in 10.0s")."
Please report at http://coq.inria.fr/bugs/.

make[3]: *** [Makefile:404: quick2vo] Error 129
make[3]: Leaving directory '/home/theo/coq/test-suite/coq-makefile/quick2vo/_test'
==========> FAILURE <==========
    coq-makefile/quick2vo/run.sh...Error!
```

Running again the failed test with `cd test-suite && rm coq-makefile/quick2vo.log && make` had it succeed.
I ran again the full test-suite `make -j4 test-suite` and this time it was successful.

**Comment by gares** [comment]
I re-read the code another time and I've a theory now. On Unix the worker is spawned using:
- pipes for the master channel
- sockets for the control channel
The control channel exists today because (on windows) the recommended way to kill a process is to ask it to commit suicide, so you need it to have a thread to listed for kind request of death. As we know threads on Caml are not very fair and sometimes they get stuck on I/O on sockets for no reason. IIRC the Caml runtime was fixed wrt this issue, but in 4.03 (or maybe 4.04).

Are the failures always on 4.02 ? @Zimmi48 your local run was on 4.02 ?

Anyway, the control socket serves no/little purpose on Unix, so we can disable it.

**Comment by Zimmi48** [comment]
That's an interesting theory, although my OCaml version was 4.06.1. Maybe there are still similar bugs in the OCaml runtime? In any case, disabling the control socket on Unix would make sense and we could observe if it makes the rate of such CI failures drop...
